Sources/DevFoundation/Event Bus/ContextualBusEventObserver.swift

```diff
@@ -0,0 +1,304 @@
+    public init(context: Context) {
+        self.dispatcher = Dispatcher(context: context)
+    }
+
+    
+    /// Adds a new handler for the specified event type.
+    ///
+    /// - Parameters:
+    ///   - eventType: The type of event to handle.
+    ///   - body: A closure to handle the event.
+    /// - Returns: An opaque object that can be used to later remove the handler.
+    @discardableResult
+    public func addHandler<Event>(
+        for eventType: Event.Type,
+        body: @escaping @Sendable (Event, inout Context) -> Void
+    ) -> AnyObject where Event: BusEvent {
+        let key = EventHandlerKey(eventType: eventType)
+        let handler = Handler<Event>(body: body)
+
+        state.withLock { (state) in
+            state.eventHandlers[key, default: []].append(handler)
+        }
+
+        return handler
+    }
+
+
+    /// Adds a new handler for the specified event type and event ID.
+    ///
+    /// - Parameters:
+    ///   - eventType: The type of event to handle.
+    ///   - id: The ID of the event to handle.
+    ///   - body: A closure to handle the event.
+    /// - Returns: An opaque object that can be used to later remove the handler.
+    @discardableResult
+    public func addHandler<Event>(
+        for eventType: Event.Type,
+        id: Event.ID,
+        body: @escaping @Sendable (Event, inout Context) -> Void
+    ) -> AnyObject where Event: BusEvent & Identifiable, Event.ID: Sendable {
+        let key = IdentifiableEventHandlerKey(eventType: eventType, eventID: id)
+        let handler = Handler<Event>(body: body)
+
+        state.withLock { (state) in
+            state.identifiableEventHandlers[key, default: []].append(handler)
+        }
+
+        return handler
+    }
+
+    
+    /// Removes the specified handler.
+    ///
+    /// Does nothing if the handler was not previously added.
+    ///
+    /// - Parameter handler: The objected returned by ``addHandler(for:body:)`` or ``addHandler(for:id:body:)`` when the
+    ///   handler was added.
+    public func removeHandler(_ handler: AnyObject) {
+        guard let handler = handler as? any EventHandler else {
+            return
+        }
+
+        state.withLock { state in
+            state.eventHandlers[EventHandlerKey(handler: handler)]?.removeAll { $0 === handler }
+
+            if let identiablekey = IdentifiableEventHandlerKey(handler: handler) {
+                state.identifiableEventHandlers[identiablekey]?.removeAll { $0 === handler }
+            }
+        }
+    }
+
+
+    public func observe<Event>(_ event: Event) where Event: BusEvent {
+        let key = EventHandlerKey(eventType: Event.self)
+        let handlers = state.withLock { $0.eventHandlers[key] as? [Handler<Event>] ?? [] }
+
+        guard !handlers.isEmpty else {
+            return
+        }
+
+        Task {
+            await dispatcher.dispatch(event, to: handlers)
+        }
```

Comment:
Should we provide any sort of guarantee that events handlers will be executed in the order the events are posted? That could matter for some use cases, I think.

Reply:
That’s how this was originally implemented (with dispatch queues). I think there will only be a problem if the task priorities are different. Perhaps I can set the task priority to something explicit so that they are all the same.

Reply:
See [this Swift forums post](https://forums.swift.org/t/actors-that-serialise-file-access/66652/17) for a little more info.

Reply:
I updated this to use a dispatch queue instead.